### PR TITLE
Fixes #23: Tracking spider attributes

### LINKED ISSUE #23: Spider attributes
So this goes along with the mention of untracked dependencies (#14), but I have a spider that has a `page_number` attribute that's incremented after every `yield self.next_page()` call.  It's initialized to 0, so when recording a callback might request page 10, but during playback the same code now requests page 0.

There's a solution to this, which is to pass `page_number` to the callback via `meta`, and since there's no branching (the increment always happens in one callback which is requested by the callback that did the previous increment) this works and isn't too difficult.

But I wonder if autounit could track spider attributes?

### PR DESCRIPTION
**Pull request by mabelvj**
Fixes #23 
## Changes

- Added tracking of the spider arguments at the input and output of processing the response, including `spider_args_in` and `spider_args_out` in `data` at `AutounitMiddleware.process_spider_output`.

- Added assertions to in `utils.test_generator` to check the spider attributes.

- There were a lot of warnings about the process opened by Popen. Changed `subprocess.Popen` + `proc.wait` for [`subprocess.run`](https://python.readthedocs.io/en/stable/library/subprocess.html#subprocess.run), which already waits for the process to stop.  Added code to emulate `subprocess.run` in python2.7.

- Updated the temporary folder structure, copying `middleware.py` and `utils.py` to the project's temporary folder where the spider files are located. This way the test is performed not over the installed package but over the local files, so it can be tested locally. We could add a test to check the installed version.

	- Updated the importing path of the test consequently and also the spider files:

		-  `scrapy_autounit.AutounitMiddleware` -> `myproject.middleware.AutounitMiddleware`

	    - `.utils` -> `myproject.utils` 

- Also added a check for the number of tests run and check if the result. If no tests run or `'OK'` not found at the end of the test message -> the test failed.

- Updated `itertree` to have a clearer visualization of the tree structure, output example:

	```shell
	tmppcpux7oo/
	    autounit/
	        __init__.py
	        tests/
	            __init__.py
	            __pycache__/
	                __init__.cpython-36.pyc
	            myspider/
	                __init__.py
	                parse/
	                    test_fixture1.py
	                    fixture1.bin
	                    __init__.py
	                    __pycache__/
	                        test_fixture1.cpython-36.pyc
	                        __init__.cpython-36.pyc
	                __pycache__/
	                    __init__.cpython-36.pyc
	        __pycache__/
	            __init__.cpython-36.pyc
	    myproject/
	        __init__.py
	        utils.py
	        settings.py
	        middleware.py
	        spiders/
	            __init__.py
	            myspider.py
	            __pycache__/
	                myspider.cpython-36.pyc
	                __init__.cpython-36.pyc
	        __pycache__/
	            middleware.cpython-36.pyc
	            settings.cpython-36.pyc
	            utils.cpython-36.pyc
	            __init__.cpython-36.pyc
	```
	Absolute path could also be added to the top-level folder.

### CODE REVIEW

File: scrapy_autounit/middleware.py [hunk 1]
```diff
@@ -74,12 +74,18 @@ def process_spider_output(self, response, result, spider):
...
+        }
 
         data = {
             'spider_name': spider.name,
             'request': request,
             'response': input_data['response'],
+            'spider_args_out': spider_attr_out,
             'result': processed_result,
+            'spider_args_in': input_data['spider_args'],
             'spider_args': input_data['spider_args'],
```
Comment (fcanobrash) [inline scrapy_autounit/middleware.py]: Looks like `spider_args` and `spider_args_in` have the same purpose.
`spider_args_in` makes more sense though since we'll have a `spider_args_out` now.
Maybe we can remove the `spider_args` key/value here and replace it with `spider_args_in` wherever it's used.

File: scrapy_autounit/utils.py [hunk 2]
```diff
@@ -149,10 +148,9 @@ def parse_request(request, spider):
     for key, value in _request.get('meta').items():
         if key != '_autounit':
             _meta[key] = parse_object(value, spider)
-    _request['meta'] = _meta
 
+    _request['meta'] = _meta.copy()
```
Comment (fcanobrash) [inline scrapy_autounit/utils.py]: Not sure if this is necessary since `_meta` is a recently created dictionary.

File: tests/test_record.py [hunk 3]
```diff
@@ -23,17 +24,46 @@ def start_requests(self):
...
+            raise
+        retcode = process.poll()
+        if check and retcode:
+            raise subprocess.CalledProcessError(
+                retcode, process.args, output=stdout, stderr=stderr)
+        return {
+            'returncode': retcode,
+            'stdout': stdout,
+            'stderr': stderr,
+        }
```
Comment (fcanobrash) [inline tests/test_record.py]: The improvements to the tests here are awesome but it would be good to have them in a separate PR just not to have them mixed with changes that are more relevant to the issue that the PR tries to solve.

**Review by fcanobrash — CHANGES_REQUESTED** [review]
@mabelvj many thanks for this PR!
Just left some comments to keep the PR more concise to its purpose.
It would be great to have only the relevant changes that help solving the issue #23 and separate the other improvements in another PR. Many thanks again 👍

File: tests/test_record.py (hunk 3 above)
Comment (mabelvj) [reply tests/test_record.py]: ok! I added this part (l41-l66) to emulate subprocess.run for python2.7, if not CI would raise an error for python2.7.  @fcanobrash

File: tests/test_record.py [hunk 4]
```diff
@@ -23,17 +24,46 @@ def start_requests(self):
...
 def run(*pargs, **kwargs):
-    proc = subprocess.Popen(*pargs, **kwargs)
-    proc.wait()
-    return {
-        'returncode': proc.returncode,
-        'stdout': proc.stdout.read(),
-        'stderr': proc.stderr.read(),
-    }
+    try:
+        process = subprocess.run(*pargs, **kwargs)
```
Comment (mabelvj) [inline tests/test_record.py]: I had to switch to subprocess.run since the previous implementation raised lots of warnings about processes not being closed.

File: scrapy_autounit/utils.py (hunk 2 above)
Comment (mabelvj) [reply scrapy_autounit/utils.py]: ok. It must be a leftover from all the debugging I had to do. I tend overkill with dicts since they sometimes give problems.

File: scrapy_autounit/middleware.py (hunk 1 above)
Comment (mabelvj) [reply scrapy_autounit/middleware.py]: Yes, I didn't want to remove it until I heared your thougths.

**Comment by mabelvj** [comment]
Hi @fcanobrash! Thanks for your comments! I will update the PR. I know not all the code is related to the issue, but for understanding the code I had to develop some code and I thought it would be a good improvement.

File: tests/test_record.py (hunk 3 above)
Comment (fcanobrash) [reply tests/test_record.py]: @mabelvj thanks for looking into the warnings issue.
They seem to be resources not closed and more specifically, looks like the issue is that we're calling `proc.stdout.read()` and never close that resource.
Closing `proc.stdout` afterwards or using a context manager around the `read()` calls should solve the problem. The same goes for `proc.stderr`.
Doing this, maybe we could still use Popen since works in both Py2 and Py3 and we'll be using less code.

File: tests/test_record.py (hunk 4 above)
Comment (fcanobrash) [reply tests/test_record.py]: Thanks for looking into this. I thing this comment is related to the same https://github.com/scrapinghub/scrapy-autounit/pull/25/files/f3e1e95b309347cfccefd9cdcd307f4355acd4e3#r311111880

File: tests/test_record.py (hunk 3 above)
Comment (fcanobrash) [reply tests/test_record.py]: By the way @mabelvj, I think this might be a simpler approach to resolve the resource warnings issue https://github.com/scrapinghub/scrapy-autounit/pull/40/files.
What do you think about that PR's approach?

Comment (mabelvj) [reply tests/test_record.py]: Thanks for the suggestion @fcanobrash. Works perfectly! Just added it.

File: tests/test_record.py [hunk 5]
```diff
@@ -155,36 +252,154 @@ def test(self):
...
+    def next_req(self):
+        self.i += 1
+        return scrapy.Request('data:text/plain,hi', dont_filter=True)
+
+    def parse(self, response):
+        if self.i < 3:
+            yield self.next_req()
+
+                """)
+
```
Comment (fcanobrash) [inline tests/test_record.py:331]: @mabelvj, many thanks for keeping up the work on this! 💪 
I think the first four cases don't actually test the spider attribute issue because they would work with the current version of autounit.
Maybe a case that tests this issue would be something like this:
```
import scrapy


class AttrsSpider(scrapy.Spider):
    name = 'attrs'

    def __init__(self, *pargs, **kwargs):
        self.page_number = 0

    def start_requests(self):
        yield scrapy.Request('data:text/plain,')

    def parse(self, response):
        self.page_number += 1
        yield {
            'page_number': self.page_number
        }
        if self.page_number < 3:
            yield scrapy.Request('data:text/plain,', dont_filter=True)
```
What do you think about creating a single test case reflecting this example, leaving the generic spider aggregate out of this PR's scope?

File: scrapy_autounit/utils.py [hunk 6]
```diff
@@ -280,14 +278,16 @@ def generate_test(fixture_path, encoding='utf-8'):
     spider_cls.update_settings(settings)
     for k, v in data.get('settings', {}).items():
         settings.set(k, v, 50)
+
     crawler = Crawler(spider_cls, settings)
-    spider = spider_cls.from_crawler(crawler, **data.get('spider_args'))
+    spider = spider_cls.from_crawler(crawler, **data.get('spider_args_in'))
```
Comment (fcanobrash) [inline scrapy_autounit/utils.py]: I just ran into an issue trying these changes out.
If you have legacy fixtures (prior to these changes) you would need to generate them all over again (because they miss `spider_args_in`).
Maybe we can look for `spider_args` first and if not present try with the new `spider_args_in`.
Also adding a `{}` as default to `get` would be nice, because no default leads to `**None` which will make python break.

**Review by fcanobrash — CHANGES_REQUESTED** [review]
These changes look great.
I think if we simplify the `test_record.py` changes to this PR scope it should be ready to merge soon.
Thanks again!

File: tests/test_record.py (hunk 5 above)
Comment (mabelvj) [reply tests/test_record.py:331]: I'm removing this test since the previous ones might already cover it, so the generic spider method can be removed.

Comment (mabelvj) [reply tests/test_record.py:331]: The version that is actually tested is the one in the files, not the installed. `middleware.py`  is copied in `_write_mw` in #L109, so we can run the test on the actual files we have in the branch without needing to install it each time.  I have inspected the internal results and they are correct. You can inspect the output setting the verbosity in `record` and `test`.

It might be out of the scope of the PR, but I developed that code to make easier the debugging and thought it might help. I can put it on a separate PR if not. What do you think?

**Review by fcanobrash — APPROVED** [review]
